### ISSUE TITLE
Error using the new cache getOrSet() method:

### ISSUE THREAD

**Issue by thyseus** [yiisoft/yii2#13399]
I want to use the new getOrSet() method of Cache: https://github.com/yiisoft/yii2/pull/13186

Unfortunately i get the following error message:

> Serialization of 'Closure' is not allowed

![image](https://cloud.githubusercontent.com/assets/654271/22018852/31c8b700-dcb1-11e6-97e6-ed4cb0572070.png)

According to stackoverflow, PHP isn´t able to do this (yet):

http://stackoverflow.com/questions/13734224/exception-serialization-of-closure-is-not-allowed

Any ideas? Or am i using the getOrSet wrong?:

```php
$offers_new = $cache->getOrSet(['merchant-offers-new', Yii::$app->user->id], function ($cache) use ($merchant) {
    return app\models\Offer::find()->leftJoin('offer_has_agency', 'offer_has_agency.offer_id = offer.id')->where([
        'merchant_id' => $merchant->id,
        'offer_has_agency.id' => null,
    ])->orderBy('created_at DESC')->all();
});
```

I am using:
```bash
$ php -v
PHP 7.0.9 (cli) (built: Jul 20 2016 11:08:23) ( ZTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
```

Thanks a lot !

**Comment by Alex-Code** [comment]
I think this is to do with trying to serialize a model object.
Does it work if you use `asArray`?

**Comment by thyseus** [comment]
Using asArray is working as expected. I think the documentation should be adjusted as long as php does not support the serialization of objects:

https://github.com/yiisoft/yii2/pull/13186/files#diff-939db7aecd14e2b72c04d8f878da53e3R552

->asArray() should be attached to the example and the situation should be explained here.

**Comment by Alex-Code** [comment]
You can serialize objects, just not closures.
I'd try a `var_dump` on one of the found records to see where the closure is.

**Comment by thyseus** [comment]
Aha. now i understand the problem. I use the the softDeleteBehavior which indeed has an closure attached to it:

```php
 ["_behaviors":"yii\base\Component":private]=>
    array(7) {
      [0]=>
      object(thyseus\files\behaviors\HasFilesBehavior)#3378 (1) {
        ["owner"]=>
        *RECURSION*
      }
      ["softDeleteBehavior"]=>
      object(yii2tech\ar\softdelete\SoftDeleteBehavior)#3379 (7) {
        ["softDeleteAttributeValues"]=>
        array(1) {
          ["deleted_at"]=>
          object(Closure)#3377 (1) {
            ["this"]=>
            *RECURSION*
          }
        }
```

**Comment by samdark** [comment]
So I guess it would fail the same way with old cache methods, right?

**Comment by thyseus** [comment]
Interestingly not. This works:

```php
$id = $merchant->id;
$offers_new = $db->cache(function($db) use($id) {
        return app\models\Offer::find()->leftJoin('offer_has_agency', 'offer_has_agency.offer_id = offer.id')->where([
            'merchant_id' => $id,
            'offer_has_agency.id' => null,
        ])->orderBy('created_at DESC')->all();
    }, 3600, $dependency);
```

**Comment by thyseus** [comment]
And doing a

`var_dump($offers_new)[0];`

shows that there also is the closure in it.

**Comment by thyseus** [comment]
Most probably you mean this attempt with 'old cache methods':

```php
$key = ['merchant-offers-new', Yii::$app->user->id];
$offers_new = $cache->get($key);

if ($offers_new === false) {
    $offers_new = app\models\Offer::find()->leftJoin('offer_has_agency', 'offer_has_agency.offer_id = offer.id')->where([
        'merchant_id' => $merchant->id,
        'offer_has_agency.id' => null,
    ])->orderBy('created_at DESC')->all();

    $cache->set($key, $offers_new);
}
```

This also fails with the Serialization of 'Closure' is not allowed error. Any ideas on how to handle it, or should we just update the documentation with an hint about this? Using ->asArray() is fine (and maybe even more performant/memory-saving than without)

**Comment by samdark** [comment]
Yes. That means it's a duplicate of old serialization issue https://github.com/yiisoft/yii2/issues/8797 which is planned for 2.1. Updating docs is a good idea.

**Comment by thyseus** [comment]
Cool! Thanks for the information.

**Comment by bologer** [comment]
@samdark is  #15410 already in 2.0.15? Because I experience the same problem with `Yii::$app->cache->getOrSet` and [this](https://github.com/yiisoft/yii2/pull/15410/commits/ef3e89e36c6137f43e4f045d5e68065f82bc6443#diff-cdc10ba7fd71c98135e6c518538b1d38L120) is exactly the line that was fixed I suppose, because I get `Closure` exception which is caused by that line

**Comment by samdark** [comment]
It is not. Will be included into 3.0.
